apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineJobRunner.java

```diff
@@ -0,0 +1,229 @@
+package gov.cms.bfd.pipeline.app;
+
+import gov.cms.bfd.pipeline.sharedutils.PipelineJob;
+import gov.cms.bfd.pipeline.sharedutils.PipelineJobOutcome;
+import gov.cms.bfd.sharedutils.interfaces.ThrowingConsumer;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Wrapper for a {@link PipelineJob} that runs the job on a schedule. Implements {@link Runnable} so
+ * it can be submitted to an {@link java.util.concurrent.ExecutorService}.
+ */
+@Slf4j
+@AllArgsConstructor
+public class PipelineJobRunner implements Runnable {
+  /** Object that tracks the status of all job runs. */
+  private final Tracker tracker;
+  /** The job we run. */
+  private final PipelineJob job;
+  /** Function used to sleep. Parameterized for use by unit tests. */
+  private final ThrowingConsumer<Long, InterruptedException> sleeper;
+  /** Used to get timestamps. Parameterized for use by unit tests. */
+  private final Clock clock;
+
+  /**
+   * Runs the job according to its schedule. If the job has no schedule simply runs the job once.
+   * Any uncaught exception thrown by the job terminates the loop. Our status is always updated in
+   * the {@link Tracker} so it knows when job runs happen as well as when the loop terminates (and
+   * for what reason).
+   */
+  @Override
+  public void run() {
+    try {
+      final var repeatMillis =
+          job.getSchedule()
+              .map(s -> Duration.of(s.getRepeatDelay(), s.getRepeatDelayUnit()))
+              .map(Duration::toMillis)
+              .orElse(0L);
+      while (tracker.jobsCanRun()) {
+        runJob();
+        if (repeatMillis <= 0 || !tracker.jobsCanRun()) {
+          break;
+        }
+        tracker.sleeping(job);
+        sleeper.accept(repeatMillis);
+      }
+      tracker.stoppingNormally(job);
+    } catch (InterruptedException ex) {
+      tracker.stoppingDueToInterrupt(job);
+    } catch (Exception ex) {
+      tracker.stoppingDueToException(job, ex);
+    } finally {
+      tracker.stopped(job);
+    }
+  }
+
+  /**
+   * Runs the job once and reports its outcome to the {@link Tracker}.
+   *
+   * @throws Exception passed through if the job terminates with an exception
+   */
+  private void runJob() throws Exception {
+    final var id = tracker.beginningRun(job);
```

Comment:
Here's a better example of where var adds ambiguity. I'm reading through the review, and now I have to ask myself what object type is this? I can go through the effort to go find what the method returns, but it seems like needless obfuscation that puts the onus on the reader to go sleuth out. (Also something easier to figure out in IDE, but hard to read in a review)

Reply:
I don't see much sleuthing required here but I'll make the change.